graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/AbstractGraphqlService.java

```diff
@@ -0,0 +1,220 @@
+                    return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT, "Missing query");
+                }
+
+                try {
+                    return executeGraphql(ctx, GraphqlRequest.of(query));
+                } catch (Exception ex) {
+                    return HttpResponse.ofFailure(ex);
+                }
+            }));
+        }
+
+        return unsupportedMediaType();
+    }
+
+    /**
+     * Handles a {@link GraphqlRequest}.
+     */
+    protected abstract HttpResponse executeGraphql(ServiceRequestContext ctx, GraphqlRequest req)
+            throws Exception;
+
+    private static Map<String, Object> toMap(@Nullable String value) throws JsonProcessingException {
+        if (Strings.isNullOrEmpty(value)) {
+            return ImmutableMap.of();
+        }
+        return parseJsonString(value);
+    }
+
+    private static Map<String, Object> toMap(@Nullable Object maybeMap) {
+        if (maybeMap == null) {
+            return ImmutableMap.of();
+        }
+
+        if (maybeMap instanceof Map) {
+            final Map<?, ?> variablesMap = (Map<?, ?>) maybeMap;
+            if (variablesMap.isEmpty()) {
+                return ImmutableMap.of();
+            }
+
+            final ImmutableMap.Builder<String, Object> builder =
+                    ImmutableMap.builderWithExpectedSize(variablesMap.size());
+            variablesMap.forEach((k, v) -> builder.put(String.valueOf(k), v));
+            return builder.build();
+        } else {
+            throw new IllegalArgumentException("Unknown parameter type variables");
+        }
+    }
+
+    private static Map<String, Object> parseJsonString(String content) throws JsonProcessingException {
+        return mapper.readValue(content, JSON_MAP);
+    }
+
+    private static HttpResponse unsupportedMediaType() {
+        return HttpResponse.of(HttpStatus.UNSUPPORTED_MEDIA_TYPE,
+                               MediaType.PLAIN_TEXT,
+                               "Unsupported media type. Only JSON compatible types and " +
+                               "application/graphql are supported.");
+    }
+
+    /**
+     * Returns the negotiated {@link MediaType}. {@link MediaType#JSON} and {@link MediaType#GRAPHQL_JSON}
+     * are commonly used for the Content-Type of a GraphQL response.
+     * If {@link HttpHeaderNames#ACCEPT} is not specified, {@link MediaType#GRAPHQL_JSON} is used by default.
+     *
+     * <p>Note that the negotiated {@link MediaType} could not be used by the implementation of
+     * {@link AbstractGraphqlService} which may choose to respond in one of several ways
+     * specified the
+     * <a href="https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#body">
+     * specification</a>.
+     */
+    @Nullable
+    private static MediaType produceType(RequestHeaders headers) {
+        // TODO(ikhoon): Add 'HttpHeaders.accept()' once https://github.com/line/armeria/pull/3714 is merged.
```

Comment:
#3714 has been merged :wink:

Reply:
Thanks! Applied in [`7d21869` (#3704)](https://github.com/line/armeria/pull/3704/commits/7d21869ce1df54d193c8cb17af6f486cd7ad8f8c). 😁